crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -428,8 +428,19 @@ impl<'this, TX: DbTxMut<'this> + DbTx<'this>> DatabaseProvider<'this, TX> {
             .map(|(id, tx)| (id, tx.into()))
             .collect::<Vec<(u64, TransactionSigned)>>();
 
-        let senders =
-            self.get_or_take::<tables::TxSenders, TAKE>(first_transaction..=last_transaction)?;
+        let senders = match self
+            .get_or_take::<tables::TxSenders, TAKE>(first_transaction..=last_transaction)
+        {
+            Ok(senders) => senders,
+            // with SenderRecovery prune part enabled, not all senders are in database anymore
+            Err(DatabaseError::Read(_)) => transactions
```

Comment:
To expand on it: if the pruning configuration deletes senders for everything but the last 10 blocks and we try to unwind 128 blocks, we will have senders for `[tip-10; tip]` blocks' transactions in the database and `[tip-128; tip-10)` will need to be recovered manually.

Reply:
What about something like this?:

```rust
        let mut senders =
            self.get_or_take::<tables::TxSenders, TAKE>(first_transaction..=last_transaction)?;

        // Create an hashset with only tx ID in senders
        let set_senders: HashSet<_> = senders.clone().into_iter().map(|(id, _)| id).collect();

        // Recover missing senders manually
        transactions.iter().for_each(|(id, tx_signed)| {
            if !set_senders.contains(id) {
                senders.push((id.to_owned(), tx_signed.recover_signer().unwrap()));
            }
        });
```

Reply:
transactions and senders come sorted, so you can just iterate through the last missing transactions

Reply:
Something like this?:
```rust
let mut tx_iter = transactions.iter();
let mut sender_iter = senders.iter().peekable();

let mut merged_senders = Vec::new();

while let Some((tx_id, tx_signed)) = tx_iter.next() {
    match sender_iter.peek() {
        Some((sender_id, _)) if sender_id == tx_id => {
            merged_senders.push(sender_iter.next().unwrap().clone());
        }
        _ => {
            merged_senders.push((*tx_id, tx_signed.recover_signer().unwrap()));
        }
    }
}
senders = merged_senders;
```

Because it's also important that the ordering remains the same and the tx IDs of `senders` and `transactions` coincide one-by-one.